nomad/state/state_store.go

```diff
@@ -4217,7 +4217,9 @@ func (s *StateSnapshot) DenormalizeAllocationSlice(allocs []*structs.Allocation)
 }
 
 // DenormalizeAllocationDiffSlice queries the Allocation for each AllocationDiff and merges
-// the updated attributes with the existing Allocation, and attaches the Job provided
+// the updated attributes with the existing Allocation, and attaches the Job provided.
+//
+// This should only be called on terminal alloc, particularly stopped or preempted allocs
 func (s *StateSnapshot) DenormalizeAllocationDiffSlice(allocDiffs []*structs.AllocationDiff) ([]*structs.Allocation, error) {
```

Comment:
It's not obvious to me how to distinguish an AllocationDiff that's meant for stopping an allocation vs non-destructive update (e.g. service name change), so cannot add an assertion here.

The distinguishing factor seems to be which field of `ApplyPlanResultsRequest` the diff appears in: e.g. `AllocsStopped` vs `AllocsPreempted` vs `AllocsUpdated`.
The distinguishing factor seems to be the field of `ApplyPlanResultsRequest` and whether the diff is